### PR TITLE
adding ability to check for smell message contents

### PR DESCRIPTION
**Pull request by tansaku**
So this is another tiny step towards https://github.com/troessner/reek/issues/589 getting set up so that we can actually test that a smell message has changed

### CODE REVIEW

**Comment by tansaku** [comment]
ah, okay, totally should have rebased first ...

File: lib/reek/smells/smell_warning.rb [hunk 1]
```diff
@@ -109,6 +109,15 @@ def common_parameters_equal?(other_parameters)
         parameters.values_at(*other_keys) == other_values
       end
 
+      def vars_equal?(other_vars)
+        match = true
+        %i(message lines context source).each do |var|
+          other_var = other_vars[var]
+          match &= (send(var) == other_var) if other_var
```
Comment (troessner) [inline lib/reek/smells/smell_warning.rb]: Please check out "common_parameters_equal?" - this is rather the way to go.

File: lib/reek/smells/smell_warning.rb [hunk 2]
```diff
@@ -57,8 +57,8 @@ def matches_smell_details?(other_smell_details = {})
         common_parameters_equal?(other_smell_details)
       end
 
-      def matches?(klass, other_smell_details = {})
-        matches_smell_type?(klass) && matches_smell_details?(other_smell_details)
+      def matches?(klass, parameters = {}, vars = {})
+        smell_classes.include?(klass.to_s) && common_parameters_equal?(parameters) && vars_equal?(vars)
```
Comment (troessner) [inline lib/reek/smells/smell_warning.rb]: That's is confusing now. The difference between "parameters" and "vars" is baroque.
I think we should rather beef up "matches_smell_details?" so that it deals with all "parameters" and "attributes" like this:

```Ruby
def matches?(klass, details = {})
  smell_classes.include?(klass.to_s) && matches_smell_details?(details)
end
```

with "details" looking like this:

```Ruby
{
  lines: [...], 
  message: ...,
  parameters: { name: ....}
}
```

And then in "matches_smell_details?" we compare those "details" in a similar, unified way similar to how we do it in "common_parameters_equal?" (which would then be obsolete).

**Comment by tansaku** [comment]
thanks for the comments @troessner - we can do this - and I've done it, but it gives me 100+ errors through the rest of the code base where many other tests have to be updated stuff like: 

```ruby
    it 'reports one-letter variable name' do
      src = 'def simple(fred) x = jim(45) end'
      expect(src).to reek_of(:UncommunicativeVariableName, name: 'x' )
    end
```

to 

```
    it 'reports one-letter variable name' do
      src = 'def simple(fred) x = jim(45) end'
      expect(src).to reek_of(:UncommunicativeVariableName, parameters: { name: 'x' })
    end
```

I suspect it might take me an hour or so to work through all fixing those ... having made the change and got it basically working I'm keen to see it merged in, but there's a lot of overhead here :-)

In the meantime here's the relevant code:

```ruby
      def matches_smell_type?(klass)
        smell_classes.include?(klass.to_s)
      end

      def matches_smell_details?(details = {})
        details_equal?(details)
      end

      def matches?(klass, details = {})
        matches_smell_type?(klass) && matches_smell_details?(details)
      end

      def details_equal?(other_details)
        detail_keys = %i(message lines context source parameters)

        other_details.keys.each do |key|
          unless detail_keys.include?(key)
            raise ArgumentError, "The detail #{key} you want to check for doesn't exist"
          end
        end

        match = true
        other_details.each do |other_key, other_value|
          match &= (send(other_key) == other_value)
        end
        match
      end
```
I'm not so happy with the length of the details equal method, but well anyway, that's a stab at it - will see if I can get all the 100+ errors fixed up and give you a green build on this ...

**Comment by troessner** [comment]
>>       def matches_smell_details?(details = {})
        details_equal?(details)
      end

      def matches?(klass, details = {})
        matches_smell_type?(klass) && matches_smell_details?(details)
      end

AFAIS the "details_equal?" method is now just an unnecessary redirection - we can kill that and just use "matches_smell_details?"

```Ruby
      def details_equal?(other_details)
        detail_keys = %i(message lines context source parameters)

        other_details.keys.each do |key|
          unless detail_keys.include?(key)
            raise ArgumentError, "The detail #{key} you want to check for doesn't exist"
          end
        end

        match = true
        other_details.each do |other_key, other_value|
          match &= (send(other_key) == other_value)
        end
        match
      end
```

Using "&=" always feels weird and is normally not necessary in Ruby.
You can just do:

```Ruby
other_details.all? do |other_key, other_value|
  send(other_key) == other_value
end
```

and get rid of the whole "match" / boolean thingie. You normally never need booleans in combination with Enumerable's in ruby.

Additionally, you need to cater to "parameters" in "details_equal?" which it doesn't seem to do anymore?

**Comment by tansaku** [comment]
thanks for those tips @troessner - I've incorporated them into the lastest code update  - see what you think

File: lib/reek/smells/smell_warning.rb [hunk 3]
```diff
@@ -17,6 +17,8 @@ class SmellWarning
       attr_reader :context, :lines, :message, :parameters, :smell_detector, :source
       def_delegators :smell_detector, :smell_category, :smell_type
 
+      DETAIL_KEYS = %i(message lines context source parameters)
```
Comment (troessner) [inline lib/reek/smells/smell_warning.rb]: "details_key" is not really expressive - how about COMPARABLE_ATTRIBUTES ?

File: lib/reek/smells/smell_warning.rb [hunk 4]
```diff
@@ -88,25 +95,23 @@ def smell_classes
         [smell_detector.smell_category, smell_detector.smell_type]
       end
 
-      def common_parameters_equal?(other_parameters)
-        other_keys   = other_parameters.keys
-        other_values = other_parameters.values
+      def check_details_exist(other_details)
+        other_details.keys.each do |key|
+          unless DETAIL_KEYS.include?(key)
+            raise ArgumentError, "The detail #{key} you want to check for doesn't exist"
```
Comment (troessner) [inline lib/reek/smells/smell_warning.rb]: Same here, let's rather use "attribute".

File: lib/reek/smells/smell_warning.rb (hunk 3 above)
Comment (troessner) [inline lib/reek/smells/smell_warning.rb]: We compare two things:

1.) "attributes": message lines context source parameters
and
2.) "parameters". This "attribute" has "keys". So let's use "keys" only in the context of "parameters", but not in the context of "attributes"

File: lib/reek/smells/smell_warning.rb [hunk 5]
```diff
@@ -53,12 +55,17 @@ def matches_smell_type?(klass)
         smell_classes.include?(klass.to_s)
       end
 
-      def matches_smell_details?(other_smell_details = {})
-        common_parameters_equal?(other_smell_details)
+      def matches_smell_details?(other_details = {})
```
Comment (troessner) [inline lib/reek/smells/smell_warning.rb]: This method is very confusing because of the naming of the methods we use in there. 
I also think I made an error when I suggested using "_details". I believe we should rather name everything here and below "_attributes", because this is in fact what we're checking for - "attributes", not "details". The wording "details" made sense before but doesn't anymore.

How about doing something like this:

```Ruby
def matches_attributes(other_attributes = {})
  return false unless common_parameters_equal?(other_attributes[:parameters])
  other_attributes.delete(:parameters)
  other_attributes.all? do |other_key, other_value|
    send(other_key) == other_value
  end
end
```

And then use "check_details_exist" only in "common_parameters_equal?" like it was before. 

```Ruby
    send(other_key) == other_value
```

will raise on invalid attributes in any case.

And then we don't need "check_details_exist" at all and you can just put the method body at the beginning of "common_parameters_equal?"

File: lib/reek/smells/smell_warning.rb (hunk 3 above)
Comment (troessner) [inline lib/reek/smells/smell_warning.rb]: Actually just see my comment directly below regarding "matches_smell_details?", that should make everthing clear ;)

File: lib/reek/smells/smell_warning.rb (hunk 4 above)
Comment (troessner) [inline lib/reek/smells/smell_warning.rb]: Actually just see my comment directly above regarding "matches_smell_details?", that should make everthing clear ;)

**Comment by tansaku** [comment]
thanks @troessner - have made some updates based on your suggestions.  Take a look.

I wasn't sure about putting the body of `check_details_exist` in with the common_parameters_equal? since that would make that method too long (and have it do two things).  Also if I remove that check from `matches_attributes?` reek flags it with feature envy ...

See how you like this latest variation

File: lib/reek/smells/smell_warning.rb [hunk 6]
```diff
@@ -53,12 +55,17 @@ def matches_smell_type?(klass)
...
+        check_attributes_comparability(other_attributes)
+        return false unless common_parameters_equal?(other_attributes[:parameters])
+        other_attributes.delete(:parameters)
+        other_attributes.all? do |other_key, other_value|
+          send(other_key) == other_value
+        end
       end
 
-      def matches?(klass, other_smell_details = {})
-        matches_smell_type?(klass) && matches_smell_details?(other_smell_details)
```
Comment (troessner) [inline lib/reek/smells/smell_warning.rb]: Let's also rename that to "attributes" here and below.

File: lib/reek/smells/smell_warning.rb [hunk 7]
```diff
@@ -88,25 +95,23 @@ def smell_classes
...
+      def check_attributes_comparability(other_attributes)
+        other_attributes.keys.each do |attr|
+          unless COMPARABLE_ATTRIBUTES.include?(attr)
+            raise ArgumentError, "The attribute #{attr} you want to check for doesn't exist"
+          end
+        end
+      end
 
+      def common_parameters_equal?(other_parameters)
+        return true unless other_parameters
```
Comment (troessner) [inline lib/reek/smells/smell_warning.rb]: This return is unnecessary (that's why it wasnt there before) and de facto a hidden type check, lets drop this line.

**Comment by troessner** [comment]
Almost there - some small changes and then lets get this merged ;)

File: spec/reek/smells/smell_warning_spec.rb [hunk 8]
```diff
@@ -87,30 +87,42 @@
   end
 
   context '#matches?' do
-    let(:uncommunicative) do
+    let(:uncomm) do
```
Comment (troessner) [inline spec/reek/smells/smell_warning_spec.rb]: Why this change? "uncomm" is a really bad name, lets revert that.

**Comment by troessner** [comment]
@tansaku please adress my latest comments and squash to one commit, then we can merge this bad boy ;)

File: spec/reek/smells/smell_warning_spec.rb (hunk 8 above)
Comment (tansaku) [inline spec/reek/smells/smell_warning_spec.rb]: this was to avoid the lines getting too long ...

File: lib/reek/smells/smell_warning.rb (hunk 7 above)
Comment (tansaku) [inline lib/reek/smells/smell_warning.rb]: I get errors if I remove this.  We have situations where we are not defining any parameters to match against - note how we are calling it:

```ruby
      def matches_attributes?(other_attributes = {})
        check_attributes_comparability(other_attributes)
        return false unless common_parameters_equal?(other_attributes[:parameters])
```

so if other_attributes has no :parameters key (which is often the case) then `common_parameters_equal?` often gets called with a nil which we need to defend against

File: lib/reek/smells/smell_warning.rb (hunk 6 above)
Comment (tansaku) [inline lib/reek/smells/smell_warning.rb]: done

**Comment by tansaku** [comment]
@troessner 2 of 3 addressed - pushing back on one (see above) - will squash once your satisfied :-)

**Comment by tansaku** [comment]
or should I squash in anticipation ...

**Comment by tansaku** [comment]
hmm, squashing much harder than I thought - complex conflicts in many more files - have made my way through them, but now stuck on two fails:

```sh

  1) Reek::Spec::ShouldReekOf smell types and smell details smell type is matching but smell details are not sets the proper error message
     Failure/Error: expect(matcher.failure_message).to\
       expected "Expected string to reek of UncommunicativeVariableName, but it didn't" to match "Expected string to reek of UncommunicativeVariableName (which it did) with smell details {:parameters=>{:name=>\"x\"}}, which it didn't"
     # ./spec/reek/spec/should_reek_of_spec.rb:110:in `block (4 levels) in <top (required)>'

  2) Reek::Spec::ShouldReekOf smell types and smell details smell type is matching but smell details are not sets the proper error message when negated
     Failure/Error: expect(matcher.failure_message_when_negated).to\
       expected "Expected string not to reek of UncommunicativeVariableName, but it did" to match "Expected string not to reek of UncommunicativeVariableName with smell details {:parameters=>{:name=>\"x\"}}, but it did"
     # ./spec/reek/spec/should_reek_of_spec.rb:118:in `block (4 levels) in <top (required)>'

Finished in 2.11 seconds (files took 1.03 seconds to load)
804 examples, 2 failures

```

which seem to relate to changes in `lib/reek/spec/should_reek_of` where we've moved from:

```ruby
      def matches?(actual)
        self.examiner = Examiner.new(actual, configuration: configuration)
        set_failure_messages
        matching_smell_types? && matching_smell_details?
      end
```

to 

```
     def matches?(actual)
        self.examiner = Examiner.new(actual, configuration: configuration)
        self.all_smells = examiner.smells
        all_smells.any? { |warning| warning.matches?(smell_category, smell_details) }
      end
```

so the failure messages are not being set in quite the same way they were before ...?

So to fix this I'm not sure which direction others are going in - are those tests that I have failing at the moment being dropped since we are going with simpler error messages that don't distinguish between smell type and associated attribute matches?

Either way I'm confused as I thought I was rebasing to master, and master on github doesn't seem to have the same updated code as I've just pulled in ...

File: lib/reek/smells/smell_warning.rb (hunk 7 above)
Comment (troessner) [inline lib/reek/smells/smell_warning.rb]: Ok, let's leave it like that.

**Comment by troessner** [comment]
Code looks good to me! Now we only need to sort out your problems above ;)
I'm on a plane the whole day tomorrow - either you figure it out on your own or I can have a more detailed look at the problem on saturday / sunday ;)

**Comment by tansaku** [comment]
thanks @troessner I've force pushed the single squashed commit so you can see precisely where we're up to.  I'm not going to be able to look at this again till Monday, so if you can see the resolution here that would be very helpful.

I believe I could resolve it either by removing the two failing tests, or by changing (reverting?) `lib/reek/spec/should_reek_of` to the state that supports those two tests, but I don't which the project wants and maybe I have misdiagnosed the issue and there is something more subtle going on. 

Your help very much appreciated

**Comment by troessner** [comment]
>> This branch has conflicts that must be resolved

You need to rebase against origin/master (which is basically the master the from this very repository), fix the conflicts and then force push.
There's a high chance that the failing tests will then be magically green ;)
If not, I can take a look.

**Comment by tansaku** [comment]
That's what I thought I did.  I'll have another look later today.  If I'm doing 'git rebase -i master' I need to do 'git fetch upstream' and git merge upstream/master first?

I thought I did do that already but will double check

**Comment by tansaku** [comment]
okay, looks like the problem was that I forget the `git merge upstream/master` step - maybe I should just do `git rebase -i upstream/master` generally.  

Anyway, green now.  Hopefully this should be ready to merge.  

Ironically even after all this, this only completes clearing my throat to get started on smell message internationalisation for https://github.com/troessner/reek/issues/589 :-)

File: spec/reek/spec/should_reek_of_spec.rb [hunk 9]
```diff
@@ -28,11 +40,16 @@
       it 'doesnt match a fragrant String' do
         expect(matcher.matches?(clean_code)).to be_falsey
       end
+
+      it 'reports the smells when should_not fails' do
+        expect(matcher.matches?(smelly_code)).to be_truthy
+        expect(matcher.failure_message_when_negated).to match('UncommunicativeVariableName')
```
Comment (troessner) [inline spec/reek/spec/should_reek_of_spec.rb]: Where is this example coming from all of a sudden? I'd rather remove this.

File: spec/reek/spec/should_reek_of_spec.rb [hunk 10]
```diff
@@ -4,21 +4,33 @@
...
+
+    it 'reports duplicate calls to @other.thing' do
+      expect(ruby).to reek_of(:Duplication, parameters: { name: '@other.thing' })
+    end
+
+    it 'reports duplicate calls to @other.thing.foo' do
+      expect(ruby).to reek_of(:Duplication, parameters: { name: '@other.thing.foo' })
+    end
+
+    it 'does not report any feature envy' do
```
Comment (troessner) [inline spec/reek/spec/should_reek_of_spec.rb]: The example description is now misleading. It's not about FeatureEnvy, it's about "does not report any other smell types" like in the old example description -> let's change that back to what it was before.

File: spec/reek/spec/should_reek_of_spec.rb [hunk 11]
```diff
@@ -4,21 +4,33 @@
...
-      expect(src).to reek_of(:Duplication)
+      expect(ruby).to reek_of(:Duplication)
     end
 
     it 'does not report any other smell types' do
-      expect(src).not_to reek_of(:FeatureEnvy)
+      expect(ruby).not_to reek_of(:FeatureEnvy)
+    end
+
+    it 'reports duplicate calls to @other.thing' do
```
Comment (troessner) [inline spec/reek/spec/should_reek_of_spec.rb]: Why are the next two examples introduced here? It seems like they are kind of copied from https://github.com/troessner/reek/blob/master/spec/reek/smells/duplicate_method_call_spec.rb ?
You can safely remove those.

File: spec/reek/smells/smell_warning_spec.rb [hunk 12]
```diff
@@ -96,12 +96,22 @@
...
 
     it 'matches on class symbol and params' do
       expect(uncommunicative.matches?(:UncommunicativeVariableName,
-                                      test: 'something')).to be_truthy
+                                      parameters: {
+                                        test: 'something'
+                                      })).to be_truthy
+    end
+
+    it 'matches on class symbol, params and vars' do
```
Comment (troessner) [inline spec/reek/smells/smell_warning_spec.rb]: "vars" is not a good name here and below (highly ambiguous). We introduced the "attributes" terminology above and should stick with it.

File: spec/reek/smells/smell_warning_spec.rb [hunk 13]
```diff
@@ -110,7 +120,15 @@
...
+                                      parameters: {
+                                        test: 'something else'
+                                      })).to be_falsy
+    end
+
+    it 'does not match on different vars' do
+      expect(uncommunicative.matches?(:UncommunicativeVariableName,
+                                      parameters: { test: 'something' },
+                                      message: 'nothing')).to be_falsy
     end
```
Comment (troessner) [inline spec/reek/smells/smell_warning_spec.rb]: SmellWarning#matches_attributes? from here https://github.com/troessner/reek/pull/732/files#diff-4c9efdbf77261fe0d68c190a1db4e3fcR58 is a very powerful method now that is not spec'ed at all it sees. Can you please add some specs?

**Comment by troessner** [comment]
I'm afraid I still can't merge this since I found some other issues in my final review, see my latest comments. Please address those and then we can merge this.;)

**Comment by tansaku** [comment]
thanks @troessner hoping to get to this soon - I already had some other additional tests for that long method that I was going to drop into a separate pull request - busy week - will try to get to this before the end of it :-)

File: spec/reek/spec/should_reek_of_spec.rb (hunk 9 above)
Comment (tansaku) [inline spec/reek/spec/should_reek_of_spec.rb]: I think it might just be due to the funky rebase I did against an old master

File: spec/reek/spec/should_reek_of_spec.rb (hunk 11 above)
Comment (tansaku) [inline spec/reek/spec/should_reek_of_spec.rb]: I think it might just be due to the funky rebase I did against an old master - sorry about that

File: spec/reek/context/module_context_spec.rb [hunk 14]
```diff
@@ -8,7 +8,7 @@
       module Fred
         def simple(x) x + 1; end
       end
-    ').to reek_of(:UncommunicativeParameterName, name: 'x')
+    ').to reek_of(:UncommunicativeParameterName, parameters: { name: 'x' })
```
Comment (mvz) [inline spec/reek/context/module_context_spec.rb]: I think we should strive for a solution that doesn't need this extra key.

File: lib/reek/smells/smell_warning.rb [hunk 15]
```diff
@@ -53,12 +55,17 @@ def matches_smell_type?(klass)
         smell_classes.include?(klass.to_s)
       end
 
-      def matches_smell_details?(other_smell_details = {})
-        common_parameters_equal?(other_smell_details)
+      def matches_attributes?(other_attributes = {})
+        check_attributes_comparability(other_attributes)
+        return false unless common_parameters_equal?(other_attributes[:parameters])
+        other_attributes.delete(:parameters)
```
Comment (mvz) [inline lib/reek/smells/smell_warning.rb]: Why not invert this logic, so you take out the values for the keys message, lines, context and source, and assume the rest are parameters? This avoids the extra indirection for the parameters.

File: lib/reek/smells/smell_warning.rb [hunk 16]
```diff
@@ -88,25 +95,23 @@ def smell_classes
...
+          end
+        end
+      end
 
+      def common_parameters_equal?(other_parameters)
+        return true unless other_parameters
+        other_keys = other_parameters.keys
         other_keys.each do |key|
-          unless parameters.key?(key)
+          unless parameters.include?(key)
```
Comment (mvz) [inline lib/reek/smells/smell_warning.rb]: Why was this changed?

File: spec/reek/spec/should_reek_of_spec.rb (hunk 9 above)
Comment (tansaku) [inline spec/reek/spec/should_reek_of_spec.rb]: investigating and I didn't write this.  It is an artifact of the funky rebase - this file had huge merge conflicts and I did my best to get it into the right shape, but because my local master was behind the upstream master it all went a bit wrong - have removed this test

File: spec/reek/spec/should_reek_of_spec.rb (hunk 10 above)
Comment (tansaku) [inline spec/reek/spec/should_reek_of_spec.rb]: Here, this is a block of tests I didn't write.  I'm just making minor changes here to fit in with the other changes you suggested.  Again I think this may be down to the funky rebase.  I think what's happened here is that the old and new versions of a test are both present, so I'm going to delete what looks like the older one that I think I accidentally pulled in via the rebase to an old master.

File: spec/reek/spec/should_reek_of_spec.rb (hunk 11 above)
Comment (tansaku) [inline spec/reek/spec/should_reek_of_spec.rb]: removing these as the diff certainly indicates I'm adding something here that isn't in the current master.  Again, I didn't write these - they were just what appeared in the file when I got massive merge conflicts with the old master

File: spec/reek/smells/smell_warning_spec.rb (hunk 12 above)
Comment (tansaku) [inline spec/reek/smells/smell_warning_spec.rb]: agreed - changed - that was an oversight not changing it retrospectively rather than an intentional insertion

File: spec/reek/smells/smell_warning_spec.rb (hunk 13 above)
Comment (tansaku) [inline spec/reek/smells/smell_warning_spec.rb]: I had a thought already about how we are not testing attribute match failure and had already written a test for that which I thought to include in a subsequent pull request.  I include that now, although that is still in the SmellWarning#matches? context.  Various of the code paths of SmellWarning#matches_attributes? are tested via the SmellWarning#matches? context.  Should they be broken out into a separate context? tested in both (no I guess) or should SmellWarning#matches_attributes? become a private method?

File: spec/reek/context/module_context_spec.rb (hunk 14 above)
Comment (tansaku) [inline spec/reek/context/module_context_spec.rb]: That sounds like a noble goal, but this is a reflection of the instance variables we have on smell_warning:

```ruby
      def initialize(smell_detector, context: '', lines: raise, message: raise,
                                     source: raise, parameters: {})
        @smell_detector = smell_detector
        @source         = source
        @context        = context.to_s
        @lines          = lines
        @message        = message
        @parameters     = parameters
      end
```

parameters is a hash - the others are not.  If we want to match against both the instance variables where one is a hash, we need to distinguish them with keys.  Maybe all the parameters should become instance variables to avoid the need for the nested parameters so to speak, but getting involved in this debate I feel a strong sense of "Yak-shaving" :-) I'm all for improving reek, but I really want to get to adding the internationalization functionality for which having the ability to check the message associated with a smell is a pre-requisite.

Maybe when making these changes is a great opportunity for a larger refactoring, but I don't feel qualified to do it.  Particularly I feel I am so far away from having an end user feature to drive the decisions that I can't really argue for or against this.

File: lib/reek/smells/smell_warning.rb (hunk 16 above)
Comment (tansaku) [inline lib/reek/smells/smell_warning.rb]: not sure - hadn't noticed the difference - have changed it back

File: lib/reek/smells/smell_warning.rb (hunk 15 above)
Comment (tansaku) [inline lib/reek/smells/smell_warning.rb]: that's an interesting idea -  but wouldn't it be cleaner in the long run to just make all the parameter options into instance variables on smell_warning? and then we don't have what seems like an artificial distinction between some aspects of the smell_warning that are instance vars, and others that are collected into a open ended group called "parameters", but as I say that feels like a large refactoring that I don't feel qualified to do ... yet :-)

Comment (troessner) [inline lib/reek/smells/smell_warning.rb]: I'm with @tansaku on this one. I believe our current structure is not quite there but we're on a good way.
I'd rather see this PR finished and merged and then address this in a follow-up.

>> and others that are collected into a open ended group called "parameters"

Exactly. I believe a cleaner, more OOP solution would be to have a class called SmellDetails and then rename the "parameters" attribute above "details" and point that to an instance. But let's leave that for a separate discussion.

File: spec/reek/context/module_context_spec.rb (hunk 14 above)
Comment (troessner) [inline spec/reek/context/module_context_spec.rb]: Agree, see above.

**Comment by troessner** [comment]
I re-reviewed this one and I am of the opinion that we should merge it now.
This: https://github.com/troessner/reek/pull/732/files#diff-4c9efdbf77261fe0d68c190a1db4e3fcR61
would be a separate refactoring that we can address later when we feel like.
@mvz any objections?

File: lib/reek/smells/smell_warning.rb (hunk 15 above)
Comment (mvz) [inline lib/reek/smells/smell_warning.rb]: Ok, let's address this in a follow-up. We just need to make sure we deal with this before the next release so we don't break the API.

**Comment by mvz** [comment]
> @mvz any objections?

No objections. LGTM!

**Comment by troessner** [comment]
Merged!
Thanks a lot @tansaku !

**Comment by tansaku** [comment]
thanks @troessner and @mvz pleased to see that merged in - now onto the real work :-)
